### DISCUSSION TITLE
How to use pretrained model from other frameworks for fine-tuning ASR?

### DISCUSSION THREAD

**Discussion by kaen2891** [speechbrain/speechbrain#1984] [Q&A]
Hello,
 
Now I have several pretrained speech models (these models are from other frameworks, i.e., [s3prl framework](https://github.com/s3prl/s3prl) or my own) and want to use them as a feature extractor for fine-tuning in speechbrain framework (i.e., [LibriSpeech](https://github.com/speechbrain/speechbrain/tree/develop/recipes/LibriSpeech) or [KoSpeech](https://github.com/speechbrain/speechbrain/tree/develop/recipes/KsponSpeech) ASR training). 

How can I address the issues below?
1) Use pretrained model for the feature extractor (weights of pretrained model will be frozen --> gradient no flow)
2) Use pretrained model for fine-tuning (weights of pretrained model will be flow. --> gradient flow)

Can you provide some reference codes or pages to handle this?

So far, I've only followed LibriSpeech ASR training with LM by the guidelines.

We desperately look forward to your response.

**Comment by kaen2891** [comment]
Could you please provide the answer to these questions? @Adel-Moumen, @TParcollet

**Comment by TParcollet** [comment]
Hi, many ways of doing this and I will only give the clean one :-p Basically, any pytorch model can be used in SB. To me, the cleanest way would be to create your own file in speechbrain.lobes.models with your custom nn.Module in it. HAve a look at the fairseq one for instance, where we load a fairseq checkpoint for w2v2. The idea is very simple: load your checkpoint with the official loader of your framework in the init function and you are done. Since you are interested about freezing and unfreezing parameters as well, going for a new class will help you achieve this in a neat way. Again, have a look at the fairseq_wav2vec file. Once your model is loaded using the tool of your framework in the init function, you can just set requires grad to False according to a parameter that you game (freeze_encoder freeze_all I dunno). OR you can even put a with torch_no_grad(): in the forward function to make sure that 0 grad is flowing after this point :-) Once you have your class, just instatiate it in the yaml like ANY other class: 

mymodel: !new:speechbrain.lobes.models.mycustom.bigmodel
    ckpt_path: path
    freeze_encoder: False

Once this is done make sure to pass mymodel to the modules in the YAML so that it is automatically wrapped to the good device and to the checkpointer in case you are fine-tuning it (so the new weights are stored!).
